src/audio/kpb.c

```diff
@@ -2378,6 +2399,190 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+	if (!last_copier_mi)
+		return -EINVAL;
+	if (outpin_idx >= KPB_MAX_SINK_CNT)
+		return -EINVAL;
+
+	int ret = 0;
+	struct comp_dev *dev = NULL;
+	struct kpb_fmt_dev_list *fmt_device_list =
+			&((struct comp_data *)comp_get_drvdata(kpb_dev))->fmt_device_list;
+
+	if (modules_to_prepare->number_of_modules != 0) {
+		fmt_device_list->kpb_list_item_[outpin_idx] = kpb_dev;
+		ret = devicelist_push(&fmt_device_list->device_list_[outpin_idx],
+				      &fmt_device_list->kpb_list_item_[outpin_idx]);
+		if (ret < 0)
+			return ret;
+	}
+	for (size_t mod_idx = 0; mod_idx < modules_to_prepare->number_of_modules; ++mod_idx) {
+		dev = get_dev_from_mi_id(modules_to_prepare->dev_ids[mod_idx].module_id,
+					 modules_to_prepare->dev_ids[mod_idx].instance_id);
+
+		if (!dev)
+			return -EINVAL;
+
+		struct comp_dev **new_list_item_ptr;
+
+		ret = alloc_fmt_module_list_item(fmt_device_list, dev, &new_list_item_ptr);
+		if (ret < 0)
+			return ret;
+		*new_list_item_ptr = dev;
+		ret = devicelist_push(&fmt_device_list->device_list_[outpin_idx],
+				      new_list_item_ptr);
+		if (ret < 0)
+			return ret;
+	}
+
+	*last_copier_mi = dev;
+
+	return ret;
+}
+
+static void clear_fmt_modules_list(struct kpb_fmt_dev_list *fmt_device_list,
+				   uint32_t outpin_idx)
+{
+	/* Note: this should be validated in layer above. */
+	assert(outpin_idx < KPB_MAX_SINK_CNT);
+	//ACE
+	devicelist_reset(&fmt_device_list->device_list_[outpin_idx], true);
+}
+
+//struct fast_mode_task fmt;
+
+static int unregister_modules_list(struct fast_mode_task *fmt,
+				   struct device_list *list_to_remove, size_t list_idx)
+{
+	if (list_to_remove == fmt->device_list_[list_idx]) {
+		fmt->device_list_[list_idx] = NULL;
+		return 0;
+	}
+	if (!fmt->device_list_[list_idx]) {
+		/* Nothing to do here */
+		return 0;
+	}
+	return -EINVAL;
+}
+
+/* Comment from ACE, may be outdated:
+ * Important: function below should be called only from within critical section
+ * (Goto KPB for more details)
+ */
+static int register_modules_list(struct fast_mode_task *fmt,
+				 struct device_list *new_list, size_t list_idx)
+{
+	if (list_idx >= ARRAY_SIZE(fmt->device_list_))
+		return -EINVAL;
+
+	/* Check if slot is free */
+	if (!fmt->device_list_[list_idx]) {
+		fmt->device_list_[list_idx] = new_list;
+		return 0;
+	}
+	if (new_list == fmt->device_list_[list_idx]) {
+		/* Already registered. */
+		return 0;
+	}
+	return -EINVAL;//was ADSP_ALREADY_IN_USE;
+}
+
+static int ConfigureFastModeTask(struct comp_dev *kpb_dev, const struct kpb_task_params *cfg,
+				 size_t pin)
+{
+	assert(cfg && pin < KPB_MAX_SINK_CNT && pin != REALTIME_PIN_ID &&
+	       cfg->dev_ids > 0);
+
+	int ret = 0;
+	/* not sure if this var is needed for anything */
+	struct comp_dev *last_copier_ptr = NULL;
+	struct comp_data *priv_data = (struct comp_data *)comp_get_drvdata(kpb_dev);
+
+	ret = unregister_modules_list(&priv_data->fmt,
+				      &priv_data->fmt_device_list.device_list_[pin],
+				      pin);
+	assert(ret == 0);
```

Comment:
same here, just supplying the same pin number twice or similar would cause an error here?